File: catalogo/sistemas/asaas/kit/typescript/cliente.ts
/**
 * Kit Asaas — cliente TypeScript (Node 18+, fetch nativo).
 *
 * Não requer edição: credenciais e ambiente vêm do .env (veja .env.example).
 *   Node 20.6+:  npx tsx --env-file=.env cliente.ts
 *   Node 18:     npm i dotenv  e rode com  npx tsx -r dotenv/config cliente.ts
 *
 * Endpoints usados (mesmos paths de endpoints.json): /v3/customers, /v3/payments, /v3/payments/{id}/pixQrCode,
 * /v3/payments/{id}/identificationField, /v3/payments/{id}/refund, /v3/subscriptions, /v3/financialTransactions,
 * /v3/finance/balance, /v3/transfers, /v3/webhooks, /v3/invoices.
 */
import { timingSafeEqual } from "node:crypto";

const URLS = { sandbox: "https://api-sandbox.asaas.com", producao: "https://api.asaas.com" } as const;
type Ambiente = keyof typeof URLS;
type Json = Record<string, any>;
type Pagina<T> = { hasMore?: boolean; totalCount?: number; data: T[] };

export class AsaasErro extends Error {
  constructor(public status: number, public erros: unknown, metodo: string, caminho: string) {
    super(`Asaas HTTP ${status} em ${metodo} ${caminho}: ${JSON.stringify(erros)}`);
  }
  get codigos(): string[] {
    return Array.isArray(this.erros) ? this.erros.map((e: any) => e?.code ?? "") : [];
  }
}

const limpar = (v?: string) => (v ?? "").trim().replace(/^['"]|['"]$/g, "");
const esperar = (s: number) => new Promise((r) => setTimeout(r, s * 1000));

export class AsaasCliente {
  readonly ambiente: Ambiente;
  readonly baseUrl: string;
  private readonly apiKey: string;
  private readonly userAgent: string;
  private readonly tentativas: number;
  private readonly timeoutMs: number;

  constructor(opts: { apiKey?: string; ambiente?: string; userAgent?: string; tentativas?: number; timeoutMs?: number } = {}) {
    this.tentativas = opts.tentativas ?? 3;
    this.timeoutMs = opts.timeoutMs ?? 30000;
    const amb = limpar(opts.ambiente ?? process.env.ASAAS_AMBIENTE).toLowerCase();
    this.ambiente = amb === "producao" ? "producao" : "sandbox"; // padrão seguro
    this.apiKey = limpar(opts.apiKey ?? process.env.ASAAS_API_KEY);
    if (!this.apiKey || this.apiKey.startsWith("<<")) {
      throw new Error("Defina ASAAS_API_KEY no .env (chave do mesmo ambiente de ASAAS_AMBIENTE).");
    }
    this.userAgent = limpar(opts.userAgent ?? process.env.ASAAS_USER_AGENT) || "kit-catalogo-typescript/1.0";
    this.baseUrl = `${URLS[this.ambiente]}/v3`;
  }

  /** Repete em 429 (sempre) e em 5xx/erro de rede só para GET. Escritas não são repetidas (sem idempotência na API). */
  async requisicao<T = Json>(metodo: string, caminho: string, opts: { params?: Json; corpo?: unknown } = {}): Promise<T> {
    const qs = opts.params
      ? "?" + new URLSearchParams(Object.entries(opts.params).filter(([, v]) => v !== undefined && v !== null && v !== "")
          .map(([k, v]) => [k, String(v)])).toString()
      : "";
    for (let tentativa = 1; tentativa <= this.tentativas; tentativa++) {
      const ultima = tentativa === this.tentativas;
      let res: Response;
      try {
        res = await fetch(this.baseUrl + caminho + qs, {
          method: metodo,
          headers: { access_token: this.apiKey, "User-Agent": this.userAgent, "Content-Type": "application/json" },
          body: opts.corpo === undefined ? undefined : JSON.stringify(opts.corpo),
          signal: AbortSignal.timeout(this.timeoutMs),
        });
      } catch (err) {
        if (metodo === "GET" && !ultima) { await esperar(2 ** tentativa); continue; }
        throw err;
      }
      if (res.status === 429 && !ultima) { await esperar(Number(res.headers.get("RateLimit-Reset") ?? 5) + 1); continue; }
      if (res.status >= 500 && metodo === "GET" && !ultima) { await esperar(2 ** tentativa); continue; }
      const texto = await res.text();
      if (!res.ok) {
        let erros: unknown = texto;
        try { erros = JSON.parse(texto).errors ?? texto; } catch { /* corpo não-JSON */ }
        throw new AsaasErro(res.status, erros, metodo, caminho);
      }
      return (texto ? JSON.parse(texto) : null) as T;
    }
    throw new Error("tentativas esgotadas");
  }

  /** Percorre offset/limit (100 por página) até hasMore=false. */
  async *paginar<T = Json>(caminho: string, filtros: Json = {}): AsyncGenerator<T> {
    for (let offset = 0; ; offset += 100) {
      const pagina = await this.requisicao<Pagina<T>>("GET", caminho, { params: { ...filtros, limit: 100, offset } });
      yield* (pagina.data ?? []);
      if (!pagina.hasMore) return;
    }
  }

  async listarTodos<T = Json>(caminho: string, filtros: Json = {}): Promise<T[]> {
    const itens: T[] = [];
    for await (const item of this.paginar<T>(caminho, filtros)) itens.push(item);
    return itens;
  }

  // ---------------------------------------------------------------- clientes
  async buscarClientePorDocumento(cpfCnpj: string): Promise<Json | undefined> {
    for await (const c of this.paginar("/customers", { cpfCnpj })) return c;
    return undefined;
  }
  criarCliente(nome: string, cpfCnpj: string, extras: Json = {}) {
    return this.requisicao("POST", "/customers", { corpo: { name: nome, cpfCnpj, ...extras } });
  }
  async obterOuCriarCliente(nome: string, cpfCnpj: string, extras: Json = {}) {
    return (await this.buscarClientePorDocumento(cpfCnpj)) ?? this.criarCliente(nome, cpfCnpj, extras);
  }

  // --------------------------------------------------------------- cobranças
  /** billingType: BOLETO | PIX | CREDIT_CARD | UNDEFINED; vencimento AAAA-MM-DD. Use externalReference. */
  criarCobranca(customer: string, billingType: string, valor: number, vencimento: string, extras: Json = {}) {
    return this.requisicao("POST", "/payments", { corpo: { customer, billingType, value: valor, dueDate: vencimento, ...extras } });
  }
  buscarCobrancasPorReferencia(externalReference: string) {
    return this.listarTodos("/payments", { externalReference });
  }
  obterCobranca(id: string) { return this.requisicao("GET", `/payments/${id}`); }
  async statusCobranca(id: string): Promise<string> { return (await this.requisicao("GET", `/payments/${id}/status`)).status; }
  /** encodedImage (PNG Base64), payload (copia-e-cola), expirationDate */
  pixQrCode(id: string) { return this.requisicao("GET", `/payments/${id}/pixQrCode`); }
  /** identificationField, nossoNumero, barCode */
  linhaDigitavel(id: string) { return this.requisicao("GET", `/payments/${id}/identificationField`); }
  excluirCobranca(id: string) { return this.requisicao("DELETE", `/payments/${id}`); }
  estornarCobranca(id: string, valor?: number, motivo?: string) {
    const corpo: Json = {};
    if (valor !== undefined) corpo.value = valor;
    if (motivo) corpo.description = motivo;
    return this.requisicao("POST", `/payments/${id}/refund`, { corpo });
  }
  listarCobrancas(filtros: Json = {}) { return this.listarTodos("/payments", filtros); }

  // ------------------------------------------------------------- assinaturas
  criarAssinatura(customer: string, billingType: string, valor: number, proximoVencimento: string, ciclo = "MONTHLY", extras: Json = {}) {
    return this.requisicao("POST", "/subscriptions", {
      corpo: { customer, billingType, value: valor, nextDueDate: proximoVencimento, cycle: ciclo, ...extras },
    });
  }
  atualizarAssinatura(id: string, campos: Json) { return this.requisicao("PUT", `/subscriptions/${id}`, { corpo: campos }); }
  cancelarAssinatura(id: string) { return this.requisicao("DELETE", `/subscriptions/${id}`); }
  cobrancasDaAssinatura(id: string) { return this.listarTodos(`/subscriptions/${id}/payments`); }

  // -------------------------------------------------------------- financeiro
  async saldo(): Promise<number> { return (await this.requisicao("GET", "/finance/balance")).balance; }
  extrato(inicio: string, fim: string) { return this.listarTodos("/financialTransactions", { startDate: inicio, finishDate: fim }); }
  /** Movimenta dinheiro; pode retornar invalid_action (autorização crítica pendente). tipoChave: CPF|CNPJ|EMAIL|PHONE|EVP */
  transferirPix(valor: number, chave: string, tipoChave: string, descricao?: string, externalReference?: string) {
    return this.requisicao("POST", "/transfers", {
      corpo: { value: valor, operationType: "PIX", pixAddressKey: chave, pixAddressKeyType: tipoChave, description: descricao, externalReference },
    });
  }
  listarTransferencias(filtros: Json = {}) { return this.listarTodos("/transfers", filtros); }
  listarNotas(filtros: Json = {}) { return this.listarTodos("/invoices", filtros); }

  // ---------------------------------------------------------------- webhooks
  criarWebhook(nome: string, url: string, email: string, authToken: string, eventos: string[], sequencial = true) {
    return this.requisicao("POST", "/webhooks", {
      corpo: { name: nome, url, email, enabled: true, interrupted: false, apiVersion: 3, authToken, events: eventos,
               sendType: sequencial ? "SEQUENTIALLY" : "NON_SEQUENTIALLY" },
    });
  }
  listarWebhooks() { return this.listarTodos("/webhooks"); }
  /** Use depois de corrigir o endpoint: reativa fila interrompida. */
  reativarWebhook(id: string) { return this.requisicao("PUT", `/webhooks/${id}`, { corpo: { interrupted: false } }); }

  /** Compara o header asaas-access-token com o token cadastrado (o Asaas não usa HMAC). */
  static webhookAutentico(headers: Record<string, string | string[] | undefined>, token = limpar(process.env.ASAAS_WEBHOOK_TOKEN)): boolean {
    const bruto = Object.entries(headers).find(([k]) => k.toLowerCase() === "asaas-access-token")?.[1];
    const recebido = Buffer.from(String(Array.isArray(bruto) ? bruto[0] : bruto ?? ""));
    const esperado = Buffer.from(token);
    return esperado.length > 0 && recebido.length === esperado.length && timingSafeEqual(recebido, esperado);
  }
}

async function exemplo() {
  const api = new AsaasCliente();
  console.log(`Ambiente: ${api.ambiente} (${api.baseUrl})`);
  if (api.ambiente === "producao") throw new Error("O exemplo cria registros de teste; use ASAAS_AMBIENTE=sandbox.");
  console.log("Saldo:", await api.saldo());
  const cliente = await api.obterOuCriarCliente("Cliente Exemplo Ltda", "00000000000191", {
    email: "financeiro@example.com", notificationDisabled: true,
  });
  const ref = "KIT-EXEMPLO-001";
  const existentes = await api.buscarCobrancasPorReferencia(ref);
  const hoje = new Date().toISOString().slice(0, 10);
  const cobranca = existentes[0] ?? (await api.criarCobranca(cliente.id, "PIX", 10, hoje, { externalReference: ref, description: "Teste do kit" }));
  const qr = await api.pixQrCode(cobranca.id);
  console.log("Cobrança:", cobranca.id, cobranca.status, "| Pix copia-e-cola:", String(qr.payload).slice(0, 30), "...");
  console.log("Cobranças no ambiente:", (await api.listarCobrancas()).length);
}

// executa o exemplo só quando o arquivo é chamado diretamente (funciona em CJS e ESM)
if (/cliente\.(ts|js|mjs|cjs)$/.test(process.argv[1] ?? "")) {
  exemplo().catch((err) => {
    console.error("Falha:", err instanceof AsaasErro ? `${err.message} | códigos: ${err.codigos.join(", ")}` : err);
    process.exit(1);
  });
}

File: catalogo/sistemas/asaas/kit/python/cliente.py
"""Kit Asaas — cliente Python (requests).

Não requer edição: credenciais e ambiente vêm do arquivo .env (veja .env.example).
    pip install requests python-dotenv
    python cliente.py          # roda o exemplo no ambiente configurado (padrão: sandbox)

Endpoints usados (mesmos paths de endpoints.json): /v3/customers, /v3/payments, /v3/payments/{id}/pixQrCode,
/v3/payments/{id}/identificationField, /v3/payments/{id}/refund, /v3/subscriptions, /v3/financialTransactions,
/v3/finance/balance, /v3/transfers, /v3/webhooks, /v3/invoices.
"""
from __future__ import annotations

import hmac
import os
import time
from typing import Any, Iterator

import requests

try:  # carrega .env se python-dotenv estiver instalado
    from dotenv import load_dotenv

    load_dotenv()
except ImportError:  # pragma: no cover
    pass

URLS = {"sandbox": "https://api-sandbox.asaas.com", "producao": "https://api.asaas.com"}


class AsaasErro(Exception):
    """Erro da API com status HTTP e a lista errors[] devolvida pelo Asaas."""

    def __init__(self, status: int, erros: Any, metodo: str, caminho: str):
        self.status, self.erros = status, erros
        super().__init__(f"Asaas HTTP {status} em {metodo} {caminho}: {erros}")

    @property
    def codigos(self) -> list[str]:
        if isinstance(self.erros, list):
            return [e.get("code", "") for e in self.erros if isinstance(e, dict)]
        return []


class AsaasCliente:
    def __init__(self, api_key: str | None = None, ambiente: str | None = None,
                 user_agent: str | None = None, timeout: int = 30, tentativas: int = 3):
        self.ambiente = (ambiente or os.getenv("ASAAS_AMBIENTE") or "sandbox").strip().lower()
        if self.ambiente not in URLS:
            self.ambiente = "sandbox"  # padrão seguro
        self.api_key = (api_key or os.getenv("ASAAS_API_KEY") or "").strip().strip("'\"")
        if not self.api_key or self.api_key.startswith("<<"):
            raise RuntimeError("Defina ASAAS_API_KEY no .env (chave do mesmo ambiente de ASAAS_AMBIENTE).")
        self.base_url = URLS[self.ambiente] + "/v3"
        self.timeout, self.tentativas = timeout, tentativas
        self.sessao = requests.Session()
        self.sessao.headers.update({
            "access_token": self.api_key,
            "User-Agent": user_agent or os.getenv("ASAAS_USER_AGENT") or "kit-catalogo-python/1.0",
            "Content-Type": "application/json",
        })

    # ------------------------------------------------------------------ base
    def requisicao(self, metodo: str, caminho: str, *, params: dict | None = None, corpo: Any = None) -> Any:
        """Repete em 429 (sempre: a chamada foi recusada) e em 5xx/erros de rede só para GET.
        Escritas não são repetidas: a API não documenta idempotência."""
        for tentativa in range(1, self.tentativas + 1):
            ultima = tentativa == self.tentativas
            try:
                r = self.sessao.request(metodo, self.base_url + caminho, params=params, json=corpo, timeout=self.timeout)
            except requests.RequestException:
                if metodo == "GET" and not ultima:
                    time.sleep(2 ** tentativa)
                    continue
                raise
            if r.status_code == 429 and not ultima:
                time.sleep(int(r.headers.get("RateLimit-Reset") or 5) + 1)
                continue
            if r.status_code >= 500 and metodo == "GET" and not ultima:
                time.sleep(2 ** tentativa)
                continue
            if r.status_code >= 400:
                try:
                    erros = r.json().get("errors", r.text)
                except ValueError:
                    erros = r.text
                raise AsaasErro(r.status_code, erros, metodo, caminho)
            return r.json() if r.content else None
        raise RuntimeError("tentativas esgotadas")

    def paginar(self, caminho: str, **filtros: Any) -> Iterator[dict]:
        """Percorre offset/limit (100 por página) até hasMore=False."""
        offset = 0
        while True:
            pagina = self.requisicao("GET", caminho, params={**filtros, "limit": 100, "offset": offset})
            yield from pagina.get("data", [])
            if not pagina.get("hasMore"):
                return
            offset += 100

    # -------------------------------------------------------------- clientes
    def buscar_cliente_por_documento(self, cpf_cnpj: str) -> dict | None:
        return next(iter(self.paginar("/customers", cpfCnpj=cpf_cnpj)), None)

    def criar_cliente(self, nome: str, cpf_cnpj: str, **extras: Any) -> dict:
        return self.requisicao("POST", "/customers", corpo={"name": nome, "cpfCnpj": cpf_cnpj, **extras})

    def obter_ou_criar_cliente(self, nome: str, cpf_cnpj: str, **extras: Any) -> dict:
        return self.buscar_cliente_por_documento(cpf_cnpj) or self.criar_cliente(nome, cpf_cnpj, **extras)

    # ------------------------------------------------------------- cobranças
    def criar_cobranca(self, customer: str, billing_type: str, valor: float, vencimento: str, **extras: Any) -> dict:
        """billing_type: BOLETO, PIX, CREDIT_CARD ou UNDEFINED. vencimento: AAAA-MM-DD.
        Informe externalReference para poder conferir duplicidade antes de repetir após timeout."""
        corpo = {"customer": customer, "billingType": billing_type, "value": valor, "dueDate": vencimento, **extras}
        return self.requisicao("POST", "/payments", corpo=corpo)

    def buscar_cobrancas_por_referencia(self, external_reference: str) -> list[dict]:
        return list(self.paginar("/payments", externalReference=external_reference))

    def obter_cobranca(self, payment_id: str) -> dict:
        return self.requisicao("GET", f"/payments/{payment_id}")

    def status_cobranca(self, payment_id: str) -> str:
        return self.requisicao("GET", f"/payments/{payment_id}/status")["status"]

    def pix_qrcode(self, payment_id: str) -> dict:
        """Retorna encodedImage (PNG Base64), payload (copia-e-cola) e expirationDate."""
        return self.requisicao("GET", f"/payments/{payment_id}/pixQrCode")

    def linha_digitavel(self, payment_id: str) -> dict:
        """Retorna identificationField, nossoNumero e barCode."""
        return self.requisicao("GET", f"/payments/{payment_id}/identificationField")

    def excluir_cobranca(self, payment_id: str) -> dict:
        return self.requisicao("DELETE", f"/payments/{payment_id}")

    def estornar_cobranca(self, payment_id: str, valor: float | None = None, motivo: str | None = None) -> dict:
        corpo: dict[str, Any] = {}
        if valor is not None:
            corpo["value"] = valor
        if motivo:
            corpo["description"] = motivo
        return self.requisicao("POST", f"/payments/{payment_id}/refund", corpo=corpo)

    def listar_cobrancas(self, **filtros: Any) -> list[dict]:
        """Ex.: listar_cobrancas(status="RECEIVED", **{"paymentDate[ge]": "2026-09-01"})"""
        return list(self.paginar("/payments", **filtros))

    # ------------------------------------------------------------ assinaturas
    def criar_assinatura(self, customer: str, billing_type: str, valor: float, proximo_vencimento: str,
                         ciclo: str = "MONTHLY", **extras: Any) -> dict:
        corpo = {"customer": customer, "billingType": billing_type, "value": valor,
                 "nextDueDate": proximo_vencimento, "cycle": ciclo, **extras}
        return self.requisicao("POST", "/subscriptions", corpo=corpo)

    def atualizar_assinatura(self, subscription_id: str, **campos: Any) -> dict:
        return self.requisicao("PUT", f"/subscriptions/{subscription_id}", corpo=campos)

    def cancelar_assinatura(self, subscription_id: str) -> dict:
        return self.requisicao("DELETE", f"/subscriptions/{subscription_id}")

    def cobrancas_da_assinatura(self, subscription_id: str) -> list[dict]:
        return list(self.paginar(f"/subscriptions/{subscription_id}/payments"))

    # ------------------------------------------------------------ financeiro
    def saldo(self) -> float:
        return self.requisicao("GET", "/finance/balance")["balance"]

    def extrato(self, inicio: str, fim: str) -> list[dict]:
        return list(self.paginar("/financialTransactions", startDate=inicio, finishDate=fim))

    def transferir_pix(self, valor: float, chave: str, tipo_chave: str, descricao: str | None = None,
                       external_reference: str | None = None) -> dict:
        """Movimenta dinheiro. tipo_chave: CPF, CNPJ, EMAIL, PHONE ou EVP. Pode exigir autorização crítica
        (erro invalid_action) se a conta não tiver whitelist de IPs ou webhook de validação de saque."""
        corpo: dict[str, Any] = {"value": valor, "operationType": "PIX", "pixAddressKey": chave, "pixAddressKeyType": tipo_chave}
        if descricao:
            corpo["description"] = descricao
        if external_reference:
            corpo["externalReference"] = external_reference
        return self.requisicao("POST", "/transfers", corpo=corpo)

    def listar_transferencias(self, **filtros: Any) -> list[dict]:
        return list(self.paginar("/transfers", **filtros))

    # ---------------------------------------------------------- notas fiscais
    def listar_notas(self, **filtros: Any) -> list[dict]:
        return list(self.paginar("/invoices", **filtros))

    # --------------------------------------------------------------- webhooks
    def criar_webhook(self, nome: str, url: str, email: str, auth_token: str, eventos: list[str],
                      sequencial: bool = True) -> dict:
        corpo = {"name": nome, "url": url, "email": email, "enabled": True, "interrupted": False,
                 "apiVersion": 3, "authToken": auth_token, "events": eventos,
                 "sendType": "SEQUENTIALLY" if sequencial else "NON_SEQUENTIALLY"}
        return self.requisicao("POST", "/webhooks", corpo=corpo)

    def listar_webhooks(self) -> list[dict]:
        return list(self.paginar("/webhooks"))

    def reativar_webhook(self, webhook_id: str) -> dict:
        """Use só depois de corrigir o endpoint: reativa uma fila interrompida."""
        return self.requisicao("PUT", f"/webhooks/{webhook_id}", corpo={"interrupted": False})

    @staticmethod
    def webhook_autentico(headers: dict, token: str | None = None) -> bool:
        """Compara o header asaas-access-token com o token cadastrado (sem HMAC no Asaas)."""
        esperado = token or os.getenv("ASAAS_WEBHOOK_TOKEN", "").strip("'\"")
        recebido = {k.lower(): v for k, v in headers.items()}.get("asaas-access-token", "")
        return bool(esperado) and hmac.compare_digest(recebido, esperado)


if __name__ == "__main__":
    api = AsaasCliente()
    print(f"Ambiente: {api.ambiente} ({api.base_url})")
    if api.ambiente == "producao":
        raise SystemExit("O exemplo cria registros de teste; rode-o apenas com ASAAS_AMBIENTE=sandbox.")
    try:
        print("Saldo:", api.saldo())
        cliente = api.obter_ou_criar_cliente("Cliente Exemplo Ltda", "00000000000191",
                                             email="financeiro@example.com", notificationDisabled=True)
        ref = "KIT-EXEMPLO-001"
        existentes = api.buscar_cobrancas_por_referencia(ref)
        cobranca = existentes[0] if existentes else api.criar_cobranca(
            cliente["id"], "PIX", 10.0, time.strftime("%Y-%m-%d"), externalReference=ref, description="Teste do kit")
        qr = api.pix_qrcode(cobranca["id"])
        print("Cobrança:", cobranca["id"], cobranca["status"], "| Pix copia-e-cola:", qr["payload"][:30], "...")
        print("Cobranças no ambiente:", len(api.listar_cobrancas()))
    except AsaasErro as erro:
        print("Falha:", erro, "| códigos:", erro.codigos)
        raise SystemExit(1)
